contracts/src/libraries/Challenge.sol

```diff
@@ -0,0 +1,26 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.17;
+
+import "./Bytes.sol";
+
+/**
+ * @title Library of challenge generation algorithm
+ * @author Seiya Kobayashi
+ */
+library Challenge {
+    /**
+     * @notice Generate a random challenge.
+     * @dev Generate a challenge of random 32 bytes (bytes32) of the given difficulty.
+     * @param _difficulty Difficulty of challenge
+     * @return challenge Generated challenge
+     */
+    function generateChallenge(
+        uint256 _difficulty
+    ) internal view returns (bytes32) {
+        return
+            Bytes.getLastNBits(
+                keccak256(abi.encodePacked(block.timestamp, msg.sender)),
```

Comment:
it's not an excellent way to use these values as randomness seeds.
Because these values are easily controlled with the cooperation of block producers.
So what about using a sequence of zeroes as a challenge like in mining algorithms?

Reply:
Add previous block hash when generating random challenge.

Reply:
Resolved in 97e43c620418aa16c65b589bf36798d0f6a43205.